src/orange/orangeinp/detail/BuildLogicUtils.hh

```diff
@@ -0,0 +1,342 @@
+//------------------------------- -*- C++ -*- -------------------------------//
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/orangeinp/detail/BuildLogicUtils.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <algorithm>
+#include <type_traits>
+#include <vector>
+
+#include "corecel/Assert.hh"
+#include "corecel/cont/VariantUtils.hh"
+#include "corecel/math/Algorithms.hh"
+#include "orange/OrangeTypes.hh"
+#include "orange/orangeinp/CsgTree.hh"
+#include "orange/orangeinp/CsgTypes.hh"
+
+namespace celeritas
+{
+namespace orangeinp
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Result of building a logic representation of a node.
+ */
+struct BuildLogicResult
+{
+    using VecLogic = std::vector<logic_int>;
+    using VecSurface = std::vector<LocalSurfaceId>;
+
+    VecSurface faces;
+    VecLogic logic;
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct a logic representation of a node.
+ *
+ * The result is a pair of vectors: the sorted surface IDs comprising the faces
+ * of this volume, and the logical representation using \em face IDs, i.e. with
+ * the surfaces remapped to the index of the surface in the face vector.
+ *
+ * The function is templated on a policy class that determines the logic
+ * representation. The policy class must have an operator() that takes a
+ * NodeId.
+ *
+ * The per-node local surfaces (faces) are sorted in ascending order of ID, not
+ * of access, since they're always evaluated sequentially rather than as part
+ * of the logic evaluation itself.
+ */
+template<class BuildLogicPolicy>
+inline BuildLogicResult build_logic(BuildLogicPolicy&& policy, NodeId n)
+{
+    static_assert(std::is_invocable_v<BuildLogicPolicy, NodeId>);
+    static_assert(std::is_rvalue_reference_v<BuildLogicPolicy&&>,
+                  "Will move from policy: rvalue ref expected");
+
+    // Construct logic vector as local surface IDs
+    auto& lgc = policy.logic();
+    CELER_EXPECT(lgc.empty());
+    policy(n);
```

Comment:
I hate to keep going with this, but the fact that `lgc` is "moved out" as a reference at the end the function, but required to be in place for `policy(n)` to work, is not intuitive and could easily cause a bug.